### PR TITLE
[Inductor CUTLASS backend] Step 3: autotune_process, and CUDABenchmarkRequest

### PR DESCRIPTION
**Pull request by ipiszy**
This is the step 3 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108015
* #107931
* __->__ #107901
* #107847
* #107802

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/autotune_process.py [hunk 1]
```diff
@@ -166,71 +168,179 @@ class BenchmarkRequest:
...
+    kernel_name: str
+    input_tensor_meta: List[TensorMeta]
+    output_tensor_meta: TensorMeta
     extra_args: Dict[str, Any]
-    num_stages: int
-    num_warps: int
 
-    input_tensors: Union["TensorMeta", List["TensorMeta"]]
-    output_tensor: Union["TensorMeta", List["TensorMeta"]]
+    def make_run_fn(
```
Comment (kadeng) [inline torch/_inductor/autotune_process.py:177]: I think this might benefit from a small doc comment telling which kind of run function it makes

File: torch/_inductor/autotune_process.py [hunk 2]
```diff
@@ -166,71 +168,179 @@ class BenchmarkRequest:
...
+        self.grid = grid
+        self.num_stages = num_stages
+        self.num_warps = num_warps
+
+    def make_run_fn(
+        self, *input_tensors: torch.Tensor, output_tensor: torch.Tensor
+    ) -> Callable[[], None]:
+        mod = PyCodeCache.load_by_key_path(self.module_cache_key, self.module_path)
+        if DEBUG:
+            print(
```
Comment (kadeng) [inline torch/_inductor/autotune_process.py:248]: Better to use logging with a DEBUG level here.

File: torch/_inductor/autotune_process.py [hunk 3]
```diff
@@ -166,71 +168,179 @@ class BenchmarkRequest:
...
+        self, *input_tensors: torch.Tensor, output_tensor: torch.Tensor
+    ) -> Callable[[], None]:
+        self.DLL, self.hash_key, self.source_file = CUDACodeCache.load(
+            self.source_code, "so"
+        )
+        args = [
+            c_void_p(tensor.data_ptr())
+            for tensor in list(input_tensors) + [output_tensor]
+        ]
+        log.debug(
```
Comment (kadeng) [inline torch/_inductor/autotune_process.py:299]: Init logger at the top of the module first?

File: torch/_inductor/autotune_process.py [hunk 4]
```diff
@@ -166,71 +168,179 @@ class BenchmarkRequest:
...
+        self.workspace_size = c_workspace_size.value
+        assert (
+            self.workspace_size == 0
+        ), "Autotune cache needs to be updated to support non-zero workspace_size!"
+        self.workspace = torch.empty(
+            self.workspace_size, dtype=torch.uint8, device="cuda"
+        )
+
+        # Generate partial function.
+        return functools.partial(
```
Comment (kadeng) [inline torch/_inductor/autotune_process.py:327]: Not saying this is a bad idea here to use functools.partial, but it can make debugging harder, as functools.partial binds it into a native object, leaving no trace in callstacks where the code is coming from. So when the function returned by functools.partial is called at a different callsite later on, it is hard to figure out which function is being called, without actually using a debugger or tracing back the origin of the function. On the other hand, it's likely slightly faster than using a python wrapper function.

File: torch/_inductor/autotune_process.py [hunk 5]
```diff
@@ -166,71 +168,179 @@ class BenchmarkRequest:
...
+            run_method,
+            *args,
+            *self.extra_args,
+            None,  # null workspace size ptr
+            c_void_p(self.workspace.data_ptr()),  # set workspace ptr
+            stream_ptr,
+        )
+
+    def cleanup_run_fn(self) -> None:
+        self.DLL.close()
```
Comment (kadeng) [inline torch/_inductor/autotune_process.py:337]: Should be guarded against being closed twice I guess.

File: torch/_inductor/utils.py [hunk 6]
```diff
@@ -46,6 +46,79 @@
...
+            return False
+        if key.startswith("void at::native::vectorized_elementwise_kernel"):
+            return False
+        return True
+
+    filtered_events = [event for event in p.key_averages() if _should_keep(event.key)]
+    log.debug("filtered events")
+    log.debug(filtered_events)
+    res = sum(event.cuda_time for event in filtered_events) / 1000.0
+    log.debug(f"profiling results: {res}")
```
Comment (kadeng) [inline torch/_inductor/utils.py:118]: I think this might warrant logging at "INFO" level.

File: torch/_inductor/autotune_process.py (hunk 2 above)
Comment (aakhundov) [reply torch/_inductor/autotune_process.py:251]: I guess, this approach is inherited from the existing code.

Comment (ipiszy) [reply torch/_inductor/autotune_process.py:251]: Yes this is from original code. I refactored it a bit so that TritonBenchmarkRequest and CUDABenchmarkRequest can share code. I'm not sure whether log works correctly when benchmark is in a subprocess so don't want to just change it. Could be a follow up task.

File: torch/_inductor/autotune_process.py (hunk 3 above)
Comment (ipiszy) [reply torch/_inductor/autotune_process.py:300]: Fixed.

File: torch/_inductor/autotune_process.py (hunk 5 above)
Comment (ipiszy) [reply torch/_inductor/autotune_process.py]: DLLWrapper.close() already handles it.

File: torch/_inductor/utils.py (hunk 6 above)
Comment (ipiszy) [reply torch/_inductor/utils.py]: Maybe end user doesn't want the benchmark logging, so still keep it as debug info.

File: torch/_inductor/utils.py [hunk 7]
```diff
@@ -45,6 +45,79 @@
...
+    for _ in range(5):
+        cache.zero_()
+        fn()
+    end_event.record()
+    torch.cuda.synchronize()
+    estimate_ms = start_event.elapsed_time(end_event) / 5
+
+    # compute number of warmup and repeat
+    n_warmup = max(1, int(warmup / estimate_ms))
+    n_repeat = max(1, int(rep / estimate_ms))
```
Comment (aakhundov) [inline torch/_inductor/utils.py:74]: I'm wondering, why not use the passed `warmup` and `rep` directly, instead of computing `n_warmup` and `n_repeat` from them?

File: torch/_inductor/utils.py [hunk 8]
```diff
@@ -45,6 +45,79 @@
...
+            return False
+        if key.startswith("void at::native::vectorized_elementwise_kernel"):
+            return False
+        return True
+
+    filtered_events = [event for event in p.key_averages() if _should_keep(event.key)]
+    log.debug("filtered events")
+    log.debug(filtered_events)
+    res = sum(event.cuda_time for event in filtered_events) / 1000.0
+    log.debug("profiling results: %s", res)
```
Comment (aakhundov) [inline torch/_inductor/utils.py:117]: Nit: add the time unit here for clarity?

File: torch/_inductor/autotune_process.py [hunk 9]
```diff
@@ -166,71 +171,197 @@ class BenchmarkRequest:
...
+            None,  # null workspace ptr
+            stream_ptr,
+        )
+        self.workspace_size = c_workspace_size.value
+        assert (
+            self.workspace_size == 0
+        ), "Autotune cache needs to be updated to support non-zero workspace_size!"
+        self.workspace = torch.empty(
+            self.workspace_size, dtype=torch.uint8, device="cuda"
+        )
```
Comment (aakhundov) [inline torch/_inductor/autotune_process.py:344]: As we're saving the `workspace` as a member now, would it make sense to skip workspace size detection and initialization if it's already done?

File: torch/_inductor/utils.py [hunk 10]
```diff
@@ -45,6 +45,79 @@
 VarRanges = Dict[sympy.Expr, sympy.Expr]
 
 
+def do_bench_using_profiling(fn: Callable[[], Any], warmup=25, rep=100) -> float:
```
Comment (jansel) [inline torch/_inductor/utils.py:48]: Why can't the existing do_bench be used?  This seems pretty similar.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/utils.py (hunk 10 above)
Comment (ipiszy) [reply torch/_inductor/utils.py:50]: do_bench also measures CPU-side overhead.

e.g. for an event sequence:
CUDA event begin -> CUDA launch kernel -> CUDA kernel execution -> CUDA event end
It measures time between CUDA event begin and CUDA event end, which contains "CUDA launch kernel" part. This part could take some CPU time. This is especially bad for CUTLASS kernels which rely on ctypes to invoke C++ functions from Python.

do_bench_using_profiling, on the other hand, relies on profiler to collect kernel device time.

File: torch/_inductor/utils.py (hunk 7 above)
Comment (ipiszy) [reply torch/_inductor/utils.py:76]: This is copied from Triton bench. I think it makes sure that there is enough time spent to warm up the device (so that the GPU frequency is set to max). If we only pass a count, for small kernels the warm up time might not be enough, while for big kernels the warm up time could be too long. n_repeat also makes sure that the total time doesn't vary a lot for small and big kernels.

**Review by ipiszy — COMMENTED** [review]
Thanks @jansel @aakhundov @kadeng , ptal~

File: test/inductor/test_inductor_utils.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        x = torch.rand(1024, 10).cuda().half()
+        w = torch.rand(512, 10).cuda().half()
+        cls._bench_fn = functools.partial(torch.nn.functional.linear, x, w)
+
+    def test_do_bench(self):
+        res = do_bench(self._bench_fn)
+        log.error("do_bench result: %s", res)
```
Comment (aakhundov) [inline test/inductor/test_inductor_utils.py:26]: Maybe use `log.warning` instead?

**Review by aakhundov — APPROVED** [review]

File: test/inductor/test_inductor_utils.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+    def test_do_bench_using_profiling(self):
+        res = do_bench_using_profiling(self._bench_fn)
+        log.error("do_bench_using_profiling result: %s", res)
```
Comment (jansel) [inline test/inductor/test_inductor_utils.py:30]: assert something about the return value

File: test/inductor/test_inductor_utils.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+class TestBench(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        x = torch.rand(1024, 10).cuda().half()
+        w = torch.rand(512, 10).cuda().half()
+        cls._bench_fn = functools.partial(torch.nn.functional.linear, x, w)
+
+    def test_do_bench(self):
+        res = do_bench(self._bench_fn)
```
Comment (jansel) [inline test/inductor/test_inductor_utils.py:25]: tests should assert something

**Review by jansel — APPROVED** [review]

File: test/inductor/test_inductor_utils.py (hunk 13 above)
Comment (ipiszy) [reply test/inductor/test_inductor_utils.py:25]: There is a check to throw exception inside `do_bench_using_profiling()`, so the test below makes sure that no exception is thrown. I also use this two tests to compare latency collected by these two functions. We could skip the `test_do_bench()` test in CI though...

**Review by kadeng — APPROVED** [review]
